File: Dave Min/2023-01-27-ValidParentheses.java
import java.util.*;
import java.io.*;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stc = new Stack<Character>();

        for(int i=0;i<s.length();i++){
            if(s.charAt(i) == ']'){
                if(stc.size() != 0 && stc.pop() == '[')  continue;
                else return false;
            }
            if(s.charAt(i) == '}'){
                if(stc.size() != 0 && stc.pop() == '{')  continue;
                else return false;
            }
            if(s.charAt(i) == ')'){
                if(stc.size() != 0 && stc.pop() == '(')  continue;
                else return false;
            }  
            stc.push(s.charAt(i));
        }

        if(stc.size() != 0){
            return false;
        }
        return true;

    }
}

File: Leo Kim/2023-01-27-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:

        d = {'(': ')', '[': ']', '{': '}'}
        stack = []

        for i in s:
            if i in d:
                stack.append(i)
            elif len(stack) == 0 or d[stack.pop()] != i:
                return False

        return len(stack) == 0

        ## O(n) time complexity but not that much fast by leetcode system...

        # while len(s) > 0:
        #     l = len(s)
        #     s = s.replace('()','').replace('{}','').replace('[]','')
        #     if l==len(s): return False
        # return True

        ## O(n^2) time complexity but.... not bad...